runtime/lua/vim/lsp/semantic_tokens.lua

```diff
@@ -656,6 +704,81 @@ function M.force_refresh(bufnr)
   highlighter:send_request()
 end
 
+--- Add a new semantic token highlighting rule.
+---
+--- @param rule function Function returning highlight name as string, or nil
+--- @param opts (nil|table) Optional keyword arguments:
+---  - desc (string, default "none"): Description of rule
+---  - bufnr (number | nil): if specified, the rule only applies in the
+---        given buffer.
+---  - ft (string | nil): if specified, applies to all files with the given
+---        filetype. Ignored if bufnr is specified. If neither is given,
+---        the rule applies globally.
+---  - pri_stack (number, default 1): how many priority levels this rule
+---        uses. Used to increase the priority for the next rule added.
+---  - priority (number | nil): use this priority for marks applied by
+---        this rule. Implies pri_stack = 0.
+---
+--- @return number id of created rule
+function M.add_rule(rule, opts)
+  opts = opts or {}
+
+  local id = next_rule_id
+  next_rule_id = next_rule_id + 1
+
+  local priority = opts.priority or highest_priority
+  local priority_delta = opts.pri_stack or opts.priority and 0 or 1
+  highest_priority = highest_priority + priority_delta
+
+  local scope = opts.bufnr and rule_scopes.buffer[opts.bufnr]
+    or opts.ft and rule_scopes.filetype[opts.ft]
+    or rule_scopes.global
+  scope[id] = { cb = rule, priority = priority, desc = (opts.desc or 'none') }
+  scope_for_rule[id] = scope
+
+  return id
+end
+
+--- Delete a semantic token highlighting rule.
+---
+--- @param id (nil|number) Rule to delete. If nil, delete all rules.
+function M.delete_rule(id)
+  local ids_to_delete = id and { id } or vim.tbl_keys(scope_for_rule)
+  for _, deleted in ipairs(ids_to_delete) do
+    scope_for_rule[deleted][deleted] = nil
+    scope_for_rule[deleted] = nil
+  end
+
+  -- If all rules are deleted, reset the implicit priority.
+  if not id then
+    highest_priority = vim.highlight.priorities.semantic_tokens
+  end
+end
+
+local default_rule = function(token, bufnr, priority, namespace)
+  local ft = vim.bo[bufnr].filetype
+
+  local mark = function(hl_group, delta)
+    vim.api.nvim_buf_set_extmark(bufnr, namespace, token.line, token.start_col, {
+      hl_group = hl_group,
+      end_col = token.end_col,
+      priority = priority + delta,
+      strict = false,
+    })
+  end
+
+  mark(string.format('@sem.type.%s.%s', token.type, ft), 0)
+
+  for _, modifier in ipairs(token.modifiers) do
+    mark(string.format('@sem.mod.%s.%s', modifier, ft), 1)
+  end
```

Comment:
That's a good point. I can add `default link` statements in `highlight_group.c`. What would be preferred:
``` vim
default link @lsp.function Function
```
or
``` vim
default link @lsp.function @function
```
My instinct is that the first is correct, but I could be convinced otherwise.

Reply:
Yes, the first option is preferable. If we make the change to separate semantic highlights from treesitter highlights, we should _completely_ separate them. (Users are still free to make the non-default link in case they have a different highlight for `@function` that they prefer.)

Reply:
Agreed.

The relative paucity of default highlight groups is a mild downside. Token types like `@lsp.type.{parameter, variable, property}` all map to `Identifier`. But, I suppose the same was true when treesitter first arrived.

Reply:
These are just fallbacks to make sure you see _something_ out of the box, no matter what colorscheme you have set.